src/Microsoft.AspNetCore.Server.IIS/IISHttpServer.cs

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO.Pipelines;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Hosting.Server;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Server.IIS
+{
+    public class IISHttpServer : IServer
+    {
+        private static NativeMethods.PFN_REQUEST_HANDLER _requestHandler = HandleRequest;
+        private static NativeMethods.PFN_SHUTDOWN_HANDLER _shutdownHandler = HandleShutdown;
+
+        private IISContextFactory _iisContextFactory;
+
+        private PipeFactory _pipeFactory = new PipeFactory();
+        private GCHandle _httpServerHandle;
+        private IApplicationLifetime _applicationLifetime;
+
+        public IFeatureCollection Features { get; } = new FeatureCollection();
+        public IISHttpServer(IApplicationLifetime applicationLifetime)
+        {
+            _applicationLifetime = applicationLifetime;
+        }
+
+        public Task StartAsync<TContext>(IHttpApplication<TContext> application, CancellationToken cancellationToken)
+        {
+            _httpServerHandle = GCHandle.Alloc(this);
+
+            _iisContextFactory = new IISContextFactory<TContext>(_pipeFactory, application);
+
+            // Start the server by registering the callback
+            // TODO the context may change here for shutdown.
+            NativeMethods.register_callbacks(_requestHandler, _shutdownHandler, (IntPtr)_httpServerHandle, (IntPtr)_httpServerHandle);
+
+            return Task.CompletedTask;
+        }
+
+        public Task StopAsync(CancellationToken cancellationToken)
+        {
+            // TODO: Drain pending requests
```

Comment:
File issues for this

Reply:
https://github.com/aspnet/IISIntegration/issues/436